cloud: add read and write metrics

adityamaru opened the pull request:
This change adds cloud.read_bytes and cloud.write_bytes metrics.
These are updated on all read/write operations to external storage endpoints.

These metrics while shared by all external storage interactions
will provide an immediate signal into whether we're seeing an
abnormal rate of reads and writes errors during support investigations.

Fixes: https://github.com/cockroachdb/cockroach/issues/89242

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/89584)
<!-- Reviewable:end -->

adityamaru commented:
A 2TB import; I paused it in the middle to fiddle with some cluster settings for funsies.

<img width="1118" alt="Screen Shot 2022-10-07 at 2 55 06 PM" src="https://user-images.githubusercontent.com/13837382/194633104-4e038e3f-991f-47cb-be86-2df5e735f109.png">

And then a backup of that data:

<img width="1005" alt="Screen Shot 2022-10-07 at 2 55 18 PM" src="https://user-images.githubusercontent.com/13837382/194633093-d99ad9d8-e931-4712-8f6d-8df2dbcd3efe.png">

dt reviewed (commented):
Are we concerned that the retry count is potentially not going to reflect retries in the implementation of some clients and/or at a higher layer either? A misleading metric might be worse than no metric eg if someone looks at it, sees it is flat and dismisses retries even if they're happening inside foo-sdk or library?

adityamaru commented:
> Are we concerned that the retry count is potentially not going to reflect retries in the implementation of some clients and/or at a higher layer either?

Yeah that's a good point and one that crossed my mind when I was plumbing counter increments into all the retry loops. We could try to be diligent about having all clients read and write paths that we retry in `pkg/cloud/<provider>`  increment this counter but maybe that's too optimistic. This doesn't address internal SDK retries or higher-level job retries either but maybe tweaking the metric description to make this clear solves that?

Alternatively, we could breakdown the retries into `cloud.resumingreader.error.retries`, `cloud.custom.error.retries` to better define what they track? The latter being any custom error handling we inject into the SDKs.

adityamaru committed:
cloud: add read and write metrics

This change adds cloud.read_bytes and cloud.write_bytes metrics.
These are updated on all read/write operations to external storage endpoints.

These metrics while shared by all external storage interactions
will provide an immediate signal into whether we're seeing an
abnormal rate of reads and writes errors during support investigations.

Fixes: #89242

Release note: None

adityamaru commented:
TFTR!

bors r=dt

stevendanna commented on pkg/roachpb/api.proto:
```diff
@@ -1464,13 +1464,6 @@ message ComputeChecksumResponse {
-  string operation = 1;
-  int32 attempt_number = 2;
-  string retry_error = 3;
-}
```
I was concerned that it wasn't safe to remove this message type in the case of importing a remote recording from a node on an older version, but it looks to me like we ignore unmarshalling errors, at least when sending them to listeners.

adityamaru replied:
Very good point but yeah https://github.com/cockroachdb/cockroach/blob/2d291751692bb40cf634504c74235ff25d1c80aa/pkg/util/tracing/crdbspan.go#L814 looks like we don't error out if unmarshalling failed. I could keep the proto around and just stop sending it for a release out of caution.

adityamaru replied:
added the proto back, will delete it in 23.2

adityamaru commented:
bors r=dt,stevendanna

adityamaru commented:
bors retry

adityamaru commented:
bors r-

adityamaru commented:
bors r+
